src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -459,40 +464,73 @@ public RestoreInfo getRestoreInfo() {
-            private Map<ShardId, ShardRestoreStatus> shards = null;
+            private final List<UpdateIndexShardRestoreStatusRequest> drainedRequests = new ArrayList<>();
+            private RestoreInfo[] batchedRestoreInfo;
+            private Map<ShardId, ShardRestoreStatus>[] batchedShards;
 
+            @SuppressWarnings("unchecked")
             @Override
             public ClusterState execute(ClusterState currentState) {
+
+                if (request.processed) {
+                    return currentState;
+                }
+
+                updatedSnapshotStateQueue.drainTo(drainedRequests);
+                
+                final int batchSize = drainedRequests.size();
+
+                // nothing to process (a previous event has processed it already)
+                if (batchSize == 0) {
+                    return currentState;
+                }
+                
                 MetaData metaData = currentState.metaData();
                 MetaData.Builder mdBuilder = MetaData.builder(currentState.metaData());
                 RestoreMetaData restore = metaData.custom(RestoreMetaData.TYPE);
                 if (restore != null) {
-                    boolean changed = false;
-                    boolean found = false;
-                    ArrayList<RestoreMetaData.Entry> entries = newArrayList();
-                    for (RestoreMetaData.Entry entry : restore.entries()) {
-                        if (entry.snapshotId().equals(request.snapshotId())) {
-                            assert !found;
-                            found = true;
-                            Map<ShardId, ShardRestoreStatus> shards = newHashMap(entry.shards());
-                            logger.trace("[{}] Updating shard [{}] with status [{}]", request.snapshotId(), request.shardId(), request.status().state());
-                            shards.put(request.shardId(), request.status());
-                            if (!completed(shards)) {
-                                entries.add(new RestoreMetaData.Entry(entry.snapshotId(), RestoreMetaData.State.STARTED, entry.indices(), ImmutableMap.copyOf(shards)));
-                            } else {
-                                logger.info("restore [{}] is done", request.snapshotId());
-                                restoreInfo = new RestoreInfo(entry.snapshotId().getSnapshot(), entry.indices(), shards.size(), shards.size() - failedShards(shards));
-                                this.shards = shards;
-                            }
-                            changed = true;
-                        } else {
-                            entries.add(entry);
-                        }
+                    int changedCount = 0;
+                    List<RestoreMetaData.Entry> entries = restore.entries();
+                    batchedRestoreInfo = new RestoreInfo[batchSize];
+                    batchedShards = new Map[batchSize];                    
+
```

Comment:
I think this can be optimized further. Here we are updating status of shards that are participating in the restore process. There are only two possible outcome of this operation for a snapshot it’s ether done when all shards are done or it is not done. It doesn’t matter if we are applying a single shard or multiple shards – there is still only one outcome per snapshot. If a snapshot is done we need to check if all shards in this snapshot has started and if they are not – create a listener. In other words instead of having an array with one element per shard it might make sense to have a map with one element per snapshot.